### ISSUE TITLE
Suggestion for the far future: iPhoneOS 1.0 app (.pxl) support

### ISSUE THREAD

**Issue by AshthorneGaming** [hikari-no-yume/touchHLE#197]
About fifteen years ago, people would write their own homebrew apps to run on the 1st generation iPod Touch and iPhone since iPhone OS 1.0 didn't have an app store. These apps would be in the .pxl format, and their structure seems to be similar to .ipa files. Some of the apps made were things like a port of the shareware of DOOM (without audio or controls), a NES emulator that may have been an early wip of Nescaline, a homebrew version of iPong, etc. I think in the far future we could maybe see support for these apps, as it's as important to see their history preserved as much as the apps made after them.

For clarification to others, there could be multiple reasons as to why iPhone OS 1.0 app support isn't mentioned anywhere here. Either it could be these apps just weren't known to the world (so far I could only find one video on YouTube showing one being used), or the fact that they rely on a completely different set of instructions, or something else. It is possible to install a .pxl file to later 32-bit iPhone models by renaming their extensions to .ipa but launching them will always result in a crash due to missing said instructions. 

For anyone who wants to see the one video of one of these apps in action, here it is: https://youtu.be/Xoz79W1bGXM?si=1NpryrpV1wo0dY46

For anyone curious, currently when renaming a .pxl to .ipa and putting it into the apps folder for touchHLE, the app will not show up in the emulator.

**Comment by hikari-no-yume** [comment]
Yeah so the thing is that iPhone OS 1.x was an unstable platform without public documentation. Apple weren't intending anyone except themselves to make apps for it. There might be some things that could be done to improve touchHLE's support for iPhone OS 1.x apps, but it's undoubtedly trickier to support than 2.x and later.

**Comment by hikari-no-yume** [comment]
By the way, touchHLE supports `.app` bundles (directories) directly. The `.ipa` file support is just for convenience and wasn't there in the first version. If you can extract the `.app` bundle from the `.pxl` file somehow, it might work in touchHLE.

**Comment by AshthorneGaming** [comment]
I've tested that just now since .pxl files can be extracted with 7zip, touchHLE just crashes when starting up.

**Comment by AshthorneGaming** [comment]
I got curious and decided to see what the latest Actions build can do with `.pxl` apps now.

Things are more stable now, which was likely unintentional but I'm glad it happened. Using `TicTacToe-0.1.pxl` as my example game, I was able to make it show up in touchHLE by extracting the file, rearranging the app files and creating folders to match the structure of an `.ipa` file, compressing it to a `.zip` archive, and renaming the extension to `.ipa`. The app won't appear if it's left uncompressed. Here is a comparison of the original `.pxl` and the altered `.ipa`.
 
<img width="1568" height="587" alt="Image" src="https://github.com/user-attachments/assets/a69c8396-0f63-4929-b6fc-b5870dca6744" />

Launching the emulator, the app appears without loading the icon or the app name. You are able to launch it, but it obviously outputs an error upon launching the app.

<img width="450" height="638" alt="Image" src="https://github.com/user-attachments/assets/7b9d37c7-7c45-4d57-97ed-10668629c8bf" /> 

<img width="484" height="265" alt="Image" src="https://github.com/user-attachments/assets/b03d340e-f6d5-4b8c-add4-78942f4cc4ef" />

In the log window, you can see that the emulator does manage to read some of the data such as the Bundle Identifier, it looks like this:

<img width="793" height="540" alt="Image" src="https://github.com/user-attachments/assets/355b2526-bc21-4e1b-87f1-7fa0ff25b230" />

<img width="1088" height="1441" alt="Image" src="https://github.com/user-attachments/assets/c3016fef-d113-4200-8cee-74cbd03ef034" />

This was only tested for educational purposes. I thought it was interesting so I figured I'd report it here.

**Comment by hikari-no-yume** [comment]
Very interesting, thanks for the report! By the looks of things, the dynamic linker in touchHLE would need some changes for these to be able to work, as the app binary is not importing the required frameworks and classes in quite the same way as all the iPhone OS 2.x and 3.x apps we've seen.

Do these .pxl apps work on iPhone OS 2.x btw? I'm wondering if Apple themselves broke compatibility here.

**Comment by AshthorneGaming** [comment]
Apparently they "work" on iPhoneOS 2.0 if the extracted files are moved to the `/applications` folder, but they suffer from stability issues and frequent crashes. Anything later is completely incompatible.
